presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,274 @@
+            }
+            if (joinDistributionType.equals(FeaturesConfig.JoinDistributionType.REPLICATED) || joinDistributionType.equals(FeaturesConfig.JoinDistributionType.AUTOMATIC)) {
+                possibleJoinNodes.add(joinNode.withDistributionType(REPLICATED));
+                possibleJoinNodes.add(joinNode.flip().withDistributionType(REPLICATED));
+            }
+
+            return planNodeOrdering().min(possibleJoinNodes);
+        }
+
+        private Ordering<PlanNode> planNodeOrdering()
+        {
+            return new Ordering<PlanNode>()
+            {
+                @Override
+                public int compare(PlanNode node1, PlanNode node2)
+                {
+                    PlanNodeCostEstimate node1Cost = lookup.getCumulativeCost(node1, session, symbolAllocator.getTypes());
+                    PlanNodeCostEstimate node2Cost = lookup.getCumulativeCost(node2, session, symbolAllocator.getTypes());
+                    return costComparator.compare(session, node1Cost, node2Cost);
+                }
+            };
+        }
+    }
+
+    /**
+     * This method generates all the ways of dividing  totalNodes into two sets
+     * each containing at least one node. It will generate one set for each
+     * possible partitioning. The other partition is implied in the absent values.
+     * In order not to generate the inverse of any set, we always include the 0th
+     * node in our sets.
+     *
+     * @param totalNodes
+     * @return A set of sets each of which defines a partitioning of totalNodes
+     */
+    public static Set<Set<Integer>> generatePartitions(int totalNodes)
+    {
+        Set<Integer> numbers = IntStream.range(0, totalNodes)
+                .boxed()
+                .collect(toImmutableSet());
+        return Sets.powerSet(numbers).stream()
+                .filter(subSet -> subSet.contains(0))
+                .filter(subSet -> subSet.size() < numbers.size())
+                .collect(toImmutableSet());
+    }
+
+    public static JoinNode createJoinAccordingToPartitioning(JoinGraphNode joinGraph, Set<Integer> leftSources, PlanNodeIdAllocator idAllocator)
+    {
+        List<PlanNode> sources = joinGraph.getSources();
+        ImmutableList.Builder<PlanNode> leftNodes = ImmutableList.builder();
+        Set<Symbol> leftSymbols = new HashSet<>();
+        ImmutableList.Builder<PlanNode> rightNodes = ImmutableList.builder();
+        for (int i = 0; i < sources.size(); i++) {
+            PlanNode source = sources.get(i);
+            if (leftSources.contains(i)) {
+                leftNodes.add(source);
+                leftSymbols.addAll(source.getOutputSymbols());
+            }
+            else {
+                rightNodes.add(source);
+            }
+        }
+
+        InnerJoinPredicateUtils.InnerJoinPushDownResult pushDownResult = sortPredicatesForInnerJoin(leftSymbols, ImmutableList.of(joinGraph.getFilter()), ImmutableList.of());
```

Comment:
> joinConjuncts and not left and rightConjuncts and the code should be simple. How? You still need to know which predicates are join predicates and which can be pushed down to either side.

I was thinking of such code (based on `PredicatePushDown.Rewriter#processInnerJoin`):
```
// create inference from all filters from all joins
allInference = createEqualityInference(allFilters);

// JOIN NODE BUILD CODE

ImmutableList.Builder<Expression> joinConjuncts = ImmutableList.builder();

// add join conjucts that were not used for inference
EqualityInference.nonInferrableConjuncts(allFilters)).stream()
  // cannot rewrite conjuct using left symbols
  .filter(conjuct -> allInference.rewriteExpression(conjuct, in(leftSymbols)) == null)
  // cannot rewrite conjuct using right symbols
  .filter(conjuct -> allInference.rewriteExpression(conjuct, not(in(leftSymbols))) == null)
  // the conjuct straddles across join sides
  .forEach(joinConjucts::add);

// scope straddling equalities get dropped in as part of the join predicate
joinConjuncts.addAll(allInference.generateEqualitiesPartitionedBy(in(leftSymbols)).getScopeStraddlingEqualities());

// NOW joinConjucts can be added as filter in newly created join
```

What do you think @rschlussel ? It seems like a simple solution. Are there some problems with it?